packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1915 @@
+};
+
+/**
+ * Describes the possible results of a {@link CaveatMutator} function.
+ */
+export enum CaveatMutatorOperation {
+  noop,
+  updateValue,
+  deleteCaveat,
+  revokePermission,
+}
+
+/**
+ * Given a caveat value, returns a {@link CaveatMutatorOperation} and, optionally,
+ * a new caveat value.
+ *
+ * @see {@link PermissionController.updatePermissionsByCaveat} for more details.
+ * @template Caveat - The caveat type for which this mutator is intended.
+ * @param caveatValue - The existing value of the caveat being mutated.
+ * @returns A tuple of the mutation result and, optionally, the new caveat
+ * value.
+ */
+export type CaveatMutator<TargetCaveat extends CaveatConstraint> = (
+  caveatValue: TargetCaveat['value'],
+) =>
+  | [Exclude<CaveatMutatorOperation, CaveatMutatorOperation.updateValue>]
+  | [CaveatMutatorOperation.updateValue, CaveatConstraint['value']];
+
+/**
+ * Extracts the permission(s) specified by the given permission and caveat
+ * specifications.
+ *
+ * @template ControllerPermissionSpecification - The permission specification(s)
+ * to extract from.
+ * @template ControllerCaveatSpecification - The caveat specification(s) to
+ * extract from. Necessary because {@link Permission} has a generic parameter
+ * that describes the allowed caveats for the permission.
+ */
+export type ExtractPermission<
+  ControllerPermissionSpecification extends PermissionSpecificationConstraint,
+  ControllerCaveatSpecification extends CaveatSpecificationConstraint,
+> = ControllerPermissionSpecification extends ValidatePermissionSpecification<ControllerPermissionSpecification>
+  ? Permission<
+      ControllerPermissionSpecification['targetKey'],
+      ExtractCaveats<ControllerCaveatSpecification>
+    >
+  : never;
+
+/**
+ * Options for the {@link PermissionController} constructor.
+ *
+ * @template ControllerPermissionSpecification - A union of the types of all
+ * permission specifications available to the controller. Any referenced caveats
+ * must be included in the controller's caveat specifications.
+ * @template ControllerCaveatSpecification - A union of the types of all
+ * caveat specifications available to the controller.
+ */
+export type PermissionControllerOptions<
+  ControllerPermissionSpecification extends PermissionSpecificationConstraint,
+  ControllerCaveatSpecification extends CaveatSpecificationConstraint,
+> = ValidatePermissionSpecification<ControllerPermissionSpecification> extends never
+  ? never
+  : {
+      messenger: PermissionControllerMessenger;
+      caveatSpecifications: CaveatSpecificationMap<ControllerCaveatSpecification>;
+      permissionSpecifications: PermissionSpecificationMap<ControllerPermissionSpecification>;
+      unrestrictedMethods: string[];
+      state?: Partial<
+        PermissionControllerState<
+          ExtractPermission<
+            ControllerPermissionSpecification,
+            ControllerCaveatSpecification
+          >
+        >
+      >;
+    };
```

Comment:
Nit: Another type simplification:
```suggestion
export type PermissionControllerOptions<
  ControllerPermissionSpecification extends ValidatePermissionSpecification<PermissionSpecificationConstraint>,
  ControllerCaveatSpecification extends CaveatSpecificationConstraint,
> = {
  messenger: PermissionControllerMessenger;
  caveatSpecifications: CaveatSpecificationMap<ControllerCaveatSpecification>;
  permissionSpecifications: PermissionSpecificationMap<ControllerPermissionSpecification>;
  unrestrictedMethods: string[];
  state?: Partial<
    PermissionControllerState<
      ExtractPermission<
        ControllerPermissionSpecification,
        ControllerCaveatSpecification
      >
    >
  >;
};
```